File: concepts/08_async_and_concurrency/05_shared_memory_atomics_and_synchronization/julia/test_02_lost_updates_and_predicates.jl
# polyglot-family: async_and_concurrency
# polyglot-concept: shared_memory_atomics_and_synchronization
# polyglot-related: languages/julia/tooling_and_runtime/13_threads_synchronization_and_memory/
# polyglot-related+: test_101_condition_variables_and_predicate_loops.jl
#
# 共同问题：read-modify-write 怎样避免丢更新；通知能否代替状态 predicate。
# 对照观察：lock 覆盖完整复合更新；Threads.Condition 与同一 lock 配合，waiter 总在循环中检查状态。

using Test
using Base.Threads

@testset "临界区和 predicate 共同定义同步协议" begin
    mutex = ReentrantLock()
    count = Ref(0)
    @sync for _ in 1:20
        Threads.@spawn @lock mutex count[] += 1
    end
    @test count[] == 20
    condition = Threads.Condition()
    ready = Ref(false)
    waiter = Threads.@spawn lock(condition) do
        while !ready[]
            wait(condition)
        end
        :ready
    end
    lock(condition) do
        ready[] = true
        notify(condition)
    end
    @test fetch(waiter) === :ready
end

File: concepts/08_async_and_concurrency/05_shared_memory_atomics_and_synchronization/rust/test_02_lost_updates_and_predicates.rs
// polyglot-family: async_and_concurrency
// polyglot-concept: shared_memory_atomics_and_synchronization
// polyglot-related: languages/rust/tests/standard_library/
// polyglot-related+: 11_threads_channels_and_memory_model/test_087_condition_variables_and_thread_local_state.rs
//
// 共同问题：read-modify-write 怎样丢更新；条件通知能否代替状态 predicate。
// 对照观察：Mutex guard covers full compound update；Condvar notification is only a wake hint, waiter loops on predicate.

use std::sync::{Arc, Condvar, Mutex};

#[test]
fn comparison() {
    let state = Arc::new((Mutex::new((false, 0)), Condvar::new()));
    let worker_state = Arc::clone(&state);
    let worker = std::thread::spawn(move || {
        let (lock, changed) = &*worker_state;
        let mut state = lock.lock().unwrap();
        state.1 += 1;
        state.0 = true;
        changed.notify_one();
    });
    let (lock, changed) = &*state;
    let mut current = lock.lock().unwrap();
    while !current.0 {
        current = changed.wait(current).unwrap();
    }
    assert_eq!(current.1, 1);
    drop(current);
    worker.join().unwrap();
}
